Add extraUrlParams and extraUrlParamsReplaceMap to amp-analytics

jmadler opened the pull request:
Implements #1881

jmadler commented:
cc @rudygalfi @avimehta @btownsend @cramforce

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+      var i = 0;
+      for (const keyrule in this.config_['customVarsMap']) {
+        var replacerule = this.config_['customVarsMap'][keyrule];
+        alert ("keyrule = " + keyrule + ";replacerule = " + replacerule);
```
I'd recommend using console.log :)

jmadler replied:
Will remove - that was for debugging

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+      // If the vendor-provided config includes a customVarsMap, apply it as a
+      // set of params to String.replace to allow aliasing of the keys in
+      // customVars.
+      var i = 0;
```
s/var/let

Rename to count

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+    if (this.config_['customVars'] && this.config_['customVarsMap']) {
+      this.config_['_customVars'] = this.config_['customVars'];
+      // If the vendor-provided config includes a customVarsMap, apply it as a
+      // set of params to String.replace to allow aliasing of the keys in
```
Nit: S

jmadler replied:
S for "// set" I presume?

cramforce replied:
Sorry, misread the text.

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+      // customVars.
+      var i = 0;
+      for (const keyrule in this.config_['customVarsMap']) {
+        var replacerule = this.config_['customVarsMap'][keyrule];
```
ReplaceMap is better, because that is what we do.

cramforce replied:
CamelCase here and elsewhere.

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+          break;
+        }
+        for (const key in this.config_['customVars']) {
+          var val = this.config_['customVars'][key];
```
Use const and let

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
+        }
+        for (const key in this.config_['customVars']) {
+          var val = this.config_['customVars'][key];
+          // only allow word chars
```
Why?

jmadler replied:
This is an instinct from Perl where you have to sanitize user-provided regexp before using it as perl regex is turing complete.  Happy to remove if you'd like.

cramforce replied:
The `.replace(string, …)` form you use above actually does not use a regex for the replace operation.

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,31 @@ export class AmpAnalytics extends AMP.BaseElement {
       return Promise.resolve();
     }
+    if (this.config_['customVars'] && this.config_['customVarsMap']) {
+      this.config_['_customVars'] = this.config_['customVars'];
```
Copy into local variable instead.

jmadler replied:
I'll remove this line entirely -- it was just to save a copy in case we might need it later

cramforce commented:
Looks great. Just some simple questions and nits.

Run `gulp lint` and `gulp presubmit` to fix the lint issues.

avimehta commented:
who are we building this for? can you point to some examples on who would use/need these?

jmadler commented:
This is for Adobe Analytics

avimehta commented:
I see. a few general concerns:

1. Having variables defined like this and then the vendor renaming them in the config looks like a weird contract between two parties. I think @cramforce already asked this and you gave an answer but why wouldn't a publisher simply specify the parameter name in their config? Isn't specifying just the parameter name and value easier, more flexible? 

  Taking a cursory look at adobe's site, it looks like the variables they use in the request are literally of the format:

  ```
prop31:en-us
prop32:en-us:adobe.com:marketing-cloud:web-analytics
prop34:9
eVar37:D=oid
eVar38:D=pid
  ``` 

  Given that, I am not sure if the `customVarsMap` part of this PR is needed. Won't just the vendor config part of your PR suffice?

2. `Custom Variables` is actually used by multiple products and it means slightly different thing in each different product. So [GA](https://developers.google.com/analytics/devguides/collection/gajs/gaTrackingCustomVariables) users might try to use this but will not see the results they would like. Same for [piwik](http://piwik.org/docs/custom-variables/) and some other products out there.

3. I will prefer a solution where the hit building/parameter config is more descriptive. For example, if we decide to do https://github.com/ampproject/amphtml/issues/1798 and https://github.com/ampproject/amphtml/issues/1701, will this work with that? To some extent, I feel that solving https://github.com/ampproject/amphtml/issues/1701 is a superset of this feature.

cramforce commented:
Maybe `customParams` is a better name, as in: It literally (with the exception of the replace) goes into the URL.

I second Avi's concern about the replace being really weird, since people could just use the output instead. But I'm happy to make things easy for Adobe customers.

jmadler commented:
1. In the case of Adobe, the key names in these key/val pairs are different in different contexts.  For example, s.evar0 is sometimes called v0.  Adobe (and other analytics providers) should be able to have that sort of flexibility in AMP Analytics.  I understand that this additional ease of use comes at a cost of reduced performance, but note that the performance impact for analytics configs not using it is a one-time check of two vars, and the entire routine is skipped if the config doesn't exist.  In terms of performance impact for those using analytics vendors that enable this routine, this O(n * m) routine runs once per page load and has a maximum "m" of 8, and an unlimited "n".  I'm happy to add a limit to "n" (customVars) to minimize the potential upper bound of performance impact in these cases.  I'd suggest 128.

2. We can rename to customParams or extraParams or whatever you like best.  I have no opinion.  Just let me know which is best for AMP Analytics.

3. #1701 and #1798 seems to me to be orthogonal to this PR.  Based on this PR, #1701 could be a config variable to omit vars and/or extraParams if the value is null; #1798 would be a config variable to specify the verb to use (POST, GET, HEAD, et al).  I'm happy to implement #1701, btw.

cramforce commented:
I'm not worried about runtime performance. Code size and complexity is the only worry.

Lets call them `extraUrlParams`

jmadler commented:
I did:
* renamed to extraUrlParams / extraUrlParamsReplaceMap
* removed debugging shims
* moved regex outside of the inner loop to the outer loop to reduce duplication of effort
* rename variables and set scope appropriately throughout (var => let/const)
* Ran `gulp lint` and `gulp presubmit` and corrected issues
I did not:
* Remove regex sanitation

PTAL

avimehta commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,37 @@ export class AmpAnalytics extends AMP.BaseElement {
+        }
+
+        // only allow word chars
+        replaceMapKey = extraUrlParamsReplaceMapKey.replace(/(\W)/g, '\$1');
```
not sure what `extraUrlParamsReplaceMapKey` is.

jmadler replied:
good catch.  corrected and tested.

avimehta commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,37 @@ export class AmpAnalytics extends AMP.BaseElement {
+      // extraUrlParams.
+      let count = 0;
+      for (let replaceMapKey in this.config_['extraUrlParamsReplaceMap']) {
+        if (++count > 16) {
```
nit: move this out into a const?

jmadler replied:
count and replaceMapKey are modified throughout the loop.

avimehta replied:
I meant 16 :)

jmadler replied:
Will do.

avimehta commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -287,6 +318,11 @@ export class AmpAnalytics extends AMP.BaseElement {
+
+    // Add any given extraUrlParams as query string params
+    if (this.config_['extraUrlParams']) {
+      request = addParamsToUrl(request, this.config_['extraUrlParams']);
```
this assumes that the parameters are appended to the url using `&` and the key and value are separated by `=`. Since that is the standard way of doing it, I am fine with it. It is probably worth noting it in docs though.

jmadler replied:
I'll add to the docs.  In fact, this PR doesn't include any docs changes, so I'll need to fix that too :)

avimehta commented:
agreed with @cramforce on code complexity. Plus I still feel that this is too specific a solution and that #1701 is a more generic replacement of this.
 
Having said that, I am still not completely sure about how this is going to be used. I am not sure there will be multiple vendor `contexts`. Is adobe going to create multiple types in vendors.js? 

Since I don't understand the usecase, I think I am okay with committing this and keeping an eye on how this will be used.

jmadler commented:
I've given this some thought and I think, more broadly, the need for the extraUrlParams is introduced because the request URLs have the variables built into the string.  

Rather than push this one through, and adding that as technical debt for later, let's grab some time to brainstorm the right approach and move forward appropriately.

cramforce commented:
I have the same questions as @avimehta, but I also want to get this resolved quickly. Main question is whether the replace list itself needs to be configured by clients or is there a single one.

I think the `extraUrlParams` feature is super useful and should definitely land.

avimehta commented:
talked offline. Iets continue with this PR. I don't think we are painting ourselves is a bad corner with this. Overall, this is a positive this for analytics and should be done. 

Jordan has some great ideas about how to make requests more flexible. Lets talk about those outside of this PR.

jmadler commented:
Final revisions made.  PTAL @cramforce before merge

cramforce commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +120,38 @@ export class AmpAnalytics extends AMP.BaseElement {
+      // of params to String.replace to allow aliasing of the keys in
+      // extraUrlParams.
+      let count = 0;
+      const MAX_REPLACES = 16;
```
This should be defined near the top of the file and come with an explanation.

cramforce commented:
One comment. Please rebase.

jmadler commented:
fixed & rebased

cramforce commented:
Rebasing appears to not have been successful. Maybe you forgot to pull down master?

jridgewell commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -120,6 +121,37 @@ export class AmpAnalytics extends AMP.BaseElement {
+        }
+
+        // only allow word chars
+        replaceMapKey = replaceMapKey.replace(/(\W)/g, '\$1');
```
What is this preventing?

jmadler replied:
early iterations treated the key as a regex.  It's now just a string, though.    I'll remove the line as it's irrelevant now.

cramforce commented:
LGTM
